### ISSUE TITLE
[SDK 52] [expo-router] React Navigation tab bars fail with "couldn't find a LinkingContext"

### ISSUE THREAD

**Issue by chriszs** [expo/expo#32371]
### Minimal reproducible example

https://github.com/chriszs/test-expo-app/tree/sdk-52-top-tabs-linkingcontext-failure

### What platform(s) does this occur on?

Android, iOS

### Where did you reproduce the issue?

in a development build

### Summary

I installed the SDK 52 beta and got the following error:

<img width="404" alt="Screenshot 2024-10-26 at 8 32 57 PM" src="https://github.com/user-attachments/assets/00f9524c-7c87-413f-81e2-4707218b1d65">

So then I reproduced it in a fresh Expo project with a minimal tabs implementation that worked in SDK 51: https://github.com/chriszs/test-expo-app/blob/sdk-52-top-tabs-linkingcontext-failure/app/(tabs)/_layout.tsx

```tsx
import type {
  ParamListBase,
  TabNavigationState,
} from "@react-navigation/native";
import { createMaterialTopTabNavigator } from "@react-navigation/material-top-tabs";
import type {
  MaterialTopTabNavigationOptions,
  MaterialTopTabNavigationEventMap,
} from "@react-navigation/material-top-tabs";
import { withLayoutContext } from "expo-router";

const { Navigator } = createMaterialTopTabNavigator();

const Tabs = withLayoutContext<
  MaterialTopTabNavigationOptions,
  typeof Navigator,
  TabNavigationState<ParamListBase>,
  MaterialTopTabNavigationEventMap
>(Navigator);

export default function TabLayout() {
  return (
    <Tabs>
      <Tabs.Screen name="index" />
    </Tabs>
  );
}
```

I have to assume something changed with how some of this works, and I see there's a new option to use headless tabs from `expo-router/ui` (#30767), but if there's a migration path I haven't found it yet.

I noticed that a linking context was added to Expo Router in #28109, as far as I can tell I have the version that should create one.

Here's it working on SDK 51 and React Navigation 6:

<img width="404" src="https://github.com/user-attachments/assets/a6b7ae4d-6e87-4f18-aa36-4c215230f6c6" />

(This isn't the only part of navigation that broke, but it is the first thing.)

### Environment

```text
expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 15.0.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.17.0 - ~/.asdf/installs/nodejs/20.17.0/bin/node
      npm: 10.8.3 - ~/.asdf/plugins/nodejs/shims/npm
      Watchman: 2024.10.21.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.0, iOS 18.0, macOS 15.0, tvOS 18.0, visionOS 2.0, watchOS 11.0
    IDEs:
      Android Studio: 2024.1 AI-241.19072.14.2412.12360217
      Xcode: 16.0/16A242d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~52.0.0-preview.5 => 52.0.0-preview.5 
      expo-router: ~4.0.0-preview.3 => 4.0.0-preview.3 
      react: 18.3.1 => 18.3.1 
      react-dom: 18.3.1 => 18.3.1 
      react-native: 0.76.0 => 0.76.0 
      react-native-web: ~0.19.13 => 0.19.13 
    Expo Workflow: bare
```

### Expo Doctor Diagnostics

```text
This project contains native project folders but also has native configuration properties in app.json, indicating it is configured to use Prebuild. When the android/ios folders are present, if you don't run prebuild in your build pipeline, the following properties will not be synced: orientation, icon, scheme, userInterfaceStyle, splash, ios, android, plugins. 

Advice: Add '/android' and '/ios' to your .gitignore file if you intend to use CNG / Prebuild. Learn more: https://docs.expo.dev/workflow/prebuild/#usage-with-eas-build
```

**Comment by michailbrynard** [comment]
I observed a similar issue with the Drawer component. I didn't have time to document it properly and have now switched back to my expo 51 build, but I saw the error:
```
 (NOBRIDGE) ERROR  Warning: Error: Couldn't find a LinkingContext context.

This error is located at:
    in DrawerItemList (created by DrawerContent)
    ...
```

This was for the following code in my _layout.tsx file which works fine on Expo 51.
```
import { Drawer } from 'expo-router/drawer';

const DrawerLayout = () => {
  return (
    <Drawer>
      <Drawer.Screen name="(tabs)" />
    </Drawer>
  );
};

export default DrawerLayout;
```

With
```
"@react-navigation/drawer": "^7.0.0-rc.32",
"expo": "^52.0.0-preview.4"
```

**Comment by lcdss** [comment]
Yesterday it was working. I updated all dependencies now and it gave me the same error.

**Comment by user-0a** [comment]
Also experiencing the same issue. Unable to use expo-router at all in SDK52.

**Comment by siddhatiwari** [comment]
I'm getting a similar error with SDK 52 related to expo-router, even with a minimal stack layout. 

The app loads fine without using expo-router as the root. 

<img width="512" alt="Screenshot 2024-10-27 at 8 23 33 AM" src="https://github.com/user-attachments/assets/ba58b96c-42e0-401f-bc67-6d01beadc724">


<img width="470" alt="Screenshot 2024-10-27 at 8 28 24 AM" src="https://github.com/user-attachments/assets/280f2a65-41e4-4e01-827b-ffdeeef5ca0d">

**Comment by chriszs** [comment]
@michailbrynard Yeah, it's also the bottom tabs in my project.

**Comment by C2thehris** [comment]
I am also seeing the "useNavigationIndependentTree" error and also the following:

Uncaught Error: A navigator can only contain 'Screen', 'Group' or 'React.Fragment' as its direct children (found 'Screen' for the screen 'index'). To render this component in the navigator, pass it in the 'component' prop to 'Screen'.
    configs useNavigationBuilder.js:72
    getRouteConfigsFromChildren useNavigationBuilder.js:44
    useNavigationBuilder useNavigationBuilder.js:140
    SlotNavigator Navigator.js:93

It seems like "useNavigationBuilder" is being passed a different Screen component than it is expecting. It expects one in node_modules/expo-router/node_modules/@react-navigation/native/node_modules/@react-navigation/core/lib/module/Screen.js

whereas its being passed one from
node_modules/@react-navigation/core/lib/module/Screen.js

**Comment by sivayapps** [comment]
I am also getting same error, unable to use Drawer navigation, 
expo SDK52
    "@react-navigation/drawer": "^7.0.0-rc.32",
    "@react-navigation/native": "^7.0.0-rc.21",
    "expo": "~52.0.0-preview.5",
Error:    
   Uncaught Error
Couldn't find a LinkingContext context.

It's very basic app generated from create-expo-app@latest --template default@beta, It's working if I remove Drawer navigation from root layout
Just started learning react-native and expo, so not sure I am making any mistake with config, but I tried many resources I could find on internet

**Comment by liontariai** [comment]
Try installing the `@react-navigation/native` version that is used in expo-router in your root package.json.
For me, i had

`"@react-navigation/native": "^6.0.2",`

installed, and the new expo-router uses 

`"@react-navigation/native": "7.0.0-rc.20",`

so I guess there was an error when resolving the package. There were two `@react-navigation/native` packages. One in the root `node_modules` and one under `expo-router/node_modules`.

**Comment by mubinansari** [comment]
> Try installing the `@react-navigation/native` version that is used in expo-router in your root package.json. For me, i had
> 
> `"@react-navigation/native": "^6.0.2",`
> 
> installed, and the new expo-router uses
> 
> `"@react-navigation/native": "7.0.0-rc.20",`
> 
> so I guess there was an error when resolving the package. There were two `@react-navigation/native` packages. One in the root `node_modules` and one under `expo-router/node_modules`.

Thanks a lot! This does fix the issue ☝🏼

**Comment by chriszs** [comment]
Can confirm downgrading to `7.0.0-rc.20` to match the `expo-router` version does seem to solve this.

**Comment by lcdss** [comment]
> Can confirm downgrading `react-navigation` deps by one version (in the case of `native` that's `7.0.0-rc.20`) to match the `expo-router` version does seem to solve this error.

I confirm the same. With all dependencies on latest version possible, just downgrading @react-navigation/native from `rc.21` to `rc.20` solved the Linking Context error.

**Comment by brentvatne** [comment]
this should be fixed with https://github.com/expo/expo/pull/32391 - install the latest expo beta version and expo-router version to get the fix!

**Comment by woodarts** [comment]
I just updated my App to latest expo version and now receiving the following error
 (NOBRIDGE) ERROR  Warning: Error: Couldn't find a screen named 'index' to use as 'initialRouteName' 

Used to work fine a day ago
Please let me know how can I fix this issue

**Comment by chriszs** [comment]
You might want to file that as a new issue, since it seems a little different and this one is closed.

**Comment by temitope-alabi-react-frontend-fcmb** [comment]
> I just updated my App to latest expo version and now receiving the following error (NOBRIDGE) ERROR Warning: Error: Couldn't find a screen named 'index' to use as 'initialRouteName'
> 
> Used to work fine a day ago Please let me know how can I fix this issue

If you want to fix this issue
try to run below command and follow the on screen instruction

npm i -g eas-cli
expo install expo@^52.0.0 --fix
npx expo-doctor@latest

then once you are done
in _layout.ts or .js or any name you called it that you initialize your <Tabs> route
 you will need to remove the initialRouteName="index". It is no more needed
 then in your app.config.js or app.json file add the key newArchEnabled: "true",
 like below
 app.config.js file
 expo: {
    newArchEnabled: "true",
    ]

Mine is working perfectly now though i have to add additional prop called  tabBarIconStyle to properly style my icon
<Tabs
      screenOptions={{
        tabBarActiveTintColor: "white",
        tabBarInactiveTintColor: "white",
        tabBarShowLabel: false,
        tabBarStyle: {
          backgroundColor: "#333333",
          borderRadius: 50,
          overflow: "hidden",
          marginHorizontal: 20,
          marginBottom: 10,
          height: 78,
          display: "flex",
          justifyContent: "space-between",
          alignItems: "center",
          flexDirection: "row",
          position: "absolute",
        },
        tabBarIconStyle: {
          alignItems: "center",
          justifyContent: "center",
          width: "100%",
          height: "100%",
        },
      }}
    > 
    </Tabs>

**Comment by marklawlor** [comment]
The originally reported issue has been resolved. If you are having an issue with similar symtoms please create a new issue and report it

**Comment by SirWilliam254** [comment]
> > I just updated my App to latest expo version and now receiving the following error (NOBRIDGE) ERROR Warning: Error: Couldn't find a screen named 'index' to use as 'initialRouteName'
> > Used to work fine a day ago Please let me know how can I fix this issue
> 
> If you want to fix this issue try to run below command and follow the on screen instruction
> 
> npm i -g eas-cli expo install expo@^52.0.0 --fix npx expo-doctor@latest
> 
> then once you are done in _layout.ts or .js or any name you called it that you initialize your route you will need to remove the initialRouteName="index". It is no more needed then in your app.config.js or app.json file add the key newArchEnabled: "true", like below app.config.js file expo: { newArchEnabled: "true", ]
> 
> Mine is working perfectly now though i have to add additional prop called tabBarIconStyle to properly style my icon <Tabs screenOptions={{ tabBarActiveTintColor: "white", tabBarInactiveTintColor: "white", tabBarShowLabel: false, tabBarStyle: { backgroundColor: "#333333", borderRadius: 50, overflow: "hidden", marginHorizontal: 20, marginBottom: 10, height: 78, display: "flex", justifyContent: "space-between", alignItems: "center", flexDirection: "row", position: "absolute", }, tabBarIconStyle: { alignItems: "center", justifyContent: "center", width: "100%", height: "100%", }, }} >

: this solves the issue, thank you.

**Comment by PavelTankovich** [comment]
+1 npx expo-doctor@latest fixed my dependencies and everything works now

**Comment by juliussneezer04** [comment]
adding as info point, `npx expo-doctor` did NOT work for me. I had to reset to versions:
```json
    "@react-navigation/native": "7.0.15",
    "expo": "52.0.46",
    "expo-router": "4.0.21",
```

Only these exact versions worked. Expo doctor suggested `^7.0.14` for `@react-navigation/native` which resolved to `7.0.14` for me - which caused this error. Same with expo `~52.0.47` was suggested which resolved to `52.0.47` but only downgrading to `52.0.46` worked for me

**Comment by jtomaszewski** [comment]
Just met this error while upgrading from expo 53 to 54.

In my case `rm yarn.lock && yarn` solve the problem.

Posting this because it took me a good hour to find the solution. (In the meantime I thought the error's at 10 other places, eh.)
